src/vm/eventpipebuffermanager.cpp

```diff
@@ -732,54 +911,70 @@ unsigned int EventPipeBufferList::GetCount() const
     return m_bufferCount;
 }
 
-EventPipeEventInstance* EventPipeBufferList::PeekNextEvent(LARGE_INTEGER beforeTimeStamp, EventPipeBuffer **pContainingBuffer)
+EventPipeBuffer* EventPipeBufferList::TryGetBuffer(LARGE_INTEGER beforeTimeStamp)
 {
-    CONTRACTL
+    LIMITED_METHOD_CONTRACT;
+    _ASSERTE(EventPipe::IsBufferManagerLockOwnedByCurrentThread());
+    /**
+     * There are 4 cases we need to handle in this function:
+     * 1) There is no buffer in the list, in this case, return nullptr
+     * 2) The head buffer is written to but not read yet, in this case, return that buffer
+     *    2.1) It is possible that the head buffer is the only buffer that is created and is empty, or
+     *    2.2) The head buffer is written to but not read
+     *    We cannot differentiate the two cases without reading it - but it is okay, in both cases, the buffer represents the head of the buffer list. 
+     *    Note that writing to the buffer can happen after we return from this function, and it is also okay.
+     * 3.) The head buffer is read but not completely reading, and
+     * 4.) The head buffer is read completely.
+     *     This case requires special attention because it is possible that the next buffer in the list contain the oldest event. Fortunately, it is 
+     *     already read so it is safe to read it to determine this case.
+     */
+
+    if (this->m_pHeadBuffer == nullptr)
     {
-        NOTHROW;
-        GC_NOTRIGGER;
-        MODE_ANY;
+        // Case 1
+        return nullptr;
     }
-    CONTRACTL_END;
-
-    // Get the current read buffer.
-    // If it's not set, start with the head buffer.
-    if(m_pReadBuffer == NULL)
+    if (this->m_pHeadBuffer->GetCreationTimeStamp().QuadPart >= beforeTimeStamp.QuadPart)
     {
-        m_pReadBuffer = m_pHeadBuffer;
+        // If the oldest buffer is still newer than the beforeTimeStamp, we can stop.
+        return nullptr;
     }
-
-    // If the read buffer is still NULL, then this list contains no buffers.
-    if(m_pReadBuffer == NULL)
+    EventPipeBufferState bufferState = this->m_pHeadBuffer->GetVolatileState();
+    if (bufferState != EventPipeBufferState::READ_ONLY)
     {
-        return NULL;
+        // Case 2 (2.1 or 2.2)
+        return this->m_pHeadBuffer;
     }
-
-    // Get the next event in the buffer.
-    EventPipeEventInstance *pNext = m_pReadBuffer->PeekNext(beforeTimeStamp);
-
-    // If the next event is NULL, then go to the next buffer.
-    if(pNext == NULL)
+    else
     {
-        m_pReadBuffer = m_pReadBuffer->GetNext();
-        if(m_pReadBuffer != NULL)
+        if (this->m_pHeadBuffer->PeekNext(beforeTimeStamp))
         {
-            pNext = m_pReadBuffer->PeekNext(beforeTimeStamp);
+            // Case 3
+            return this->m_pHeadBuffer;
+        }
+        else
+        {
+            // Case 4
+            return this->m_pHeadBuffer->GetNext();
```

Comment:
m_pHeadBuffer was older than beforeTimeStamp, but m_pHeadBuffer->GetNext() might not be.